plugins/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,299 @@
+
+			var kvPairs []datasync.ProtoWatchResp
+
+			n := 0
+			for prefix, iter := range e.GetValues() {
+				var keyVals []datasync.KeyVal
+				for x, done := iter.GetNext(); !done; x, done = iter.GetNext() {
+					kvPairs = append(kvPairs, &ProtoWatchResp{
+						Key:  x.GetKey(),
+						lazy: x,
+					})
+					keyVals = append(keyVals, x)
+					n++
+				}
+				if len(keyVals) > 0 {
+					p.Log.Debugf(" - Resync: %q (%v items)", prefix, len(keyVals))
+				} else {
+					p.Log.Debugf(" - Resync: %q", prefix)
+				}
+				for _, x := range keyVals {
+					p.Log.Debugf("\t - %q: (rev: %v)", x.GetKey(), x.GetRevision())
+				}
+			}
+			p.Log.Debugf("Resync with %d items", n)
+
+			ctx := context.Background()
+			ctx = kvs.WithResync(ctx, kvs.FullResync, true)
+			//ctx = kvs.WithRetry(ctx, time.Second, true)
+			err, _ := p.PushData(ctx, kvPairs)
+			e.Done(err)
+
+			/*n := 0
+			txn := p.KVScheduler.StartNBTransaction()
+			for prefix, iter := range e.GetValues() {
+				var keyVals []datasync.KeyVal
+				for x, done := iter.GetNext(); !done; x, done = iter.GetNext() {
+					keyVals = append(keyVals, x)
+					txn.SetValue(x.GetKey(), x)
+					n++
+				}
+				if len(keyVals) > 0 {
+					p.Log.Debugf(" - Resync: %q (%v items)", prefix, len(keyVals))
+				} else {
+					p.Log.Debugf(" - Resync: %q", prefix)
+				}
+				for _, x := range keyVals {
+					p.Log.Debugf("\t - %q: (rev: %v)", x.GetKey(), x.GetRevision())
+				}
+			}
+			p.Log.Debugf("Resyncing %d items", n)
+
+			ctx := context.Background()
+			//ctx = kvs.WithRetry(ctx, time.Second, true)
+			ctx = kvs.WithResync(ctx, kvs.FullResync, true)
+
+			kvErrs, err := txn.Commit(ctx)
+			if err != nil {
+				p.Log.Errorf("transaction failed: %v", err)
+			} else if len(kvErrs) > 0 {
+				p.Log.Warnf("transaction finished with %d errors: %+v", len(kvErrs), kvErrs)
+			} else {
+				p.Log.Infof("transaction successful")
+			}
+			e.Done(err)*/
+		}
+	}
+}
+
+func (p *Plugin) ListData() map[string]proto.Message {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.store.db
+}
+
+// PushData ...
+func (p *Plugin) PushData(ctx context.Context, kvPairs []datasync.ProtoWatchResp) (err error, kvErrs []kvs.KeyWithError) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if typ, _ := kvs.IsResync(ctx); typ == kvs.FullResync {
+		p.store.Reset()
+	}
+
+	txn := p.KVScheduler.StartNBTransaction()
+
+	for _, kv := range kvPairs {
+		p.Log.Debugf(" - %v: %q (rev: %v)",
+			kv.GetChangeType(), kv.GetKey(), kv.GetRevision())
+
+		if kv.GetChangeType() == datasync.Delete {
+			txn.SetValue(kv.GetKey(), nil)
+			p.store.Delete(kv.GetKey())
+		} else {
+			txn.SetValue(kv.GetKey(), kv)
+			p.store.Update(kv.GetKey(), kv.(*ProtoWatchResp).Val)
+		}
+	}
+
+	seqID, err := txn.Commit(ctx)
+	if err != nil {
+		p.Log.Errorf("Transaction %d failed: %v", seqID, err)
+		return err, nil
+	} else {
+		if txErr, ok := err.(*kvs.TransactionError); ok && len(txErr.GetKVErrors()) > 0 {
```

Comment:
The processing of TransactionError should be under the `if` branch above.

Reply:
My mistake. It was first made when the Commit returned two ways for errors.